src/NSubstitute/Core/CallRouterResolver.cs

```diff
@@ -1,31 +1,49 @@
-﻿using System.Collections.Concurrent;
+﻿using System;
 using NSubstitute.Exceptions;
 
 namespace NSubstitute.Core
 {
     public class CallRouterResolver : ICallRouterResolver
     {
-        readonly ConcurrentDictionary<object, ICallRouter> _callRouterMappings = new ConcurrentDictionary<object, ICallRouter>();
-
         public ICallRouter ResolveFor(object substitute)
         {
             if (substitute == null) throw new NullSubstituteReferenceException();
-            if (substitute is ICallRouter) return (ICallRouter)substitute;
-            if (substitute is ICallRouterProvider) return ((ICallRouterProvider) substitute).CallRouter;
-            ICallRouter callRouter;
-            if (_callRouterMappings.TryGetValue(substitute, out callRouter))
+
+            var callRouter = TryGetRouterFromObject(substitute);
+            if (callRouter == null && substitute is Delegate delegateProxy)
             {
-                return callRouter;
+                callRouter = TryGetRouterFromObject(delegateProxy.Target);
             }
-            throw new NotASubstituteException();
+
+            if (callRouter == null)
+                throw new NotASubstituteException();
+
+            return callRouter;
         }
 
         public void Register(object proxy, ICallRouter callRouter)
```

Comment:
I'm not sure this method is needed anymore? It was previously called from `SubstituteFactory` to ensure the router was registered for the delegate. I think it can be removed.

Reply:
I decided to keep that code for the consistency. However, you are absolutely right - it's code for the purpose of having code. I've stripped that to keep it simpler. If later we again need a registry - we'll introduce that code again.